src/shogun/lib/any.h

```diff
@@ -327,18 +327,34 @@ namespace shogun
 		template <class T, class S>
 		inline auto free_array(T* ptr, S size)
 		{
+			if (!ptr)
+			{
+				return;
+			}
 			SG_FREE(ptr);
 		}
 
 		template <class S>
 		inline auto free_array(CSGObject** ptr, S size)
 		{
+			if (!ptr)
+			{
+				return;
+			}
 			for (S i = 0; i < size; ++i)
 			{
 				ptr[i]->unref();
 			}
 			SG_FREE(ptr);
 		}
+
+		template <class T>
+		inline void copy_array(T* begin, T* end, T* dst)
+		{
+			std::copy(begin, end, dst);
+		}
+
+		void copy_array(CSGObject** begin, CSGObject** end, CSGObject** dst);
```

Comment:
There is a problem here:
This will only match if the pointer type is `CSGObject` but not for any subclass (discovered this in my own tests when writing clone). See below how to reproduce
The thing is: we decided to register the base classes directly, e.g. `CKernel`, without casting them to `CSGObject`
I think we will have to use the `decltype` trick you did for cloning single objects ....